renderdoc/driver/vulkan/vk_resources.h

```diff
@@ -1071,6 +1071,47 @@ struct MemRefs
 struct ImgRefs;
 struct ImageState;
 
+class VkPendingSubmissionCompleteCallbacks
+{
+public:
+  VkPendingSubmissionCompleteCallbacks() = default;
+  VkPendingSubmissionCompleteCallbacks(VkDevice device, VkEvent event,
+                                       rdcarray<std::function<void()>> callbacks);
+  VkPendingSubmissionCompleteCallbacks(const VkPendingSubmissionCompleteCallbacks &other);
+  VkPendingSubmissionCompleteCallbacks(VkPendingSubmissionCompleteCallbacks &&other);
+  VkPendingSubmissionCompleteCallbacks &operator=(VkPendingSubmissionCompleteCallbacks other);
+
+  ~VkPendingSubmissionCompleteCallbacks();
+
+  operator VkEvent() const
+  {
+    SCOPED_LOCK(m_Lock);
+    return d == NULL ? VK_NULL_HANDLE : d->event;
+  }
+
+  void swap(VkPendingSubmissionCompleteCallbacks &other)
+  {
+    using std::swap;
+
+    swap(callbacks, other.callbacks);
+    swap(d, other.d);
+  }
+
+  rdcarray<std::function<void()>> callbacks;
+
+private:
+  struct DPtr
+  {
+    int32_t refCount;
+    VkDevice device;
+    VkEvent event;
+  };
+
+  static Threading::CriticalSection m_Lock;
+
+  DPtr *d = NULL;
```

Comment:
Rather than having a value-copyable wrapper around a pointer like this you can remove the internal hidden pointer and make the `VkPendingSubmissionCompleteCallbacks` itself dynamically allocated. It can track its own refcount and delete itself when it reaches 0 as with `ResourceRecord`.